chain/actors/actor_storagemarket.go

```diff
@@ -0,0 +1,604 @@
+		}
+
+		if dealInfo.ActivationEpoch > 0 {
+			// this probably can't happen in practice
+			return nil, aerrors.New(3, "deal already active")
+		}
+
+		dealInfo.ActivationEpoch = vmctx.BlockHeight()
+
+		if err := deals.Set(deal, &dealInfo); err != nil {
+			return nil, aerrors.HandleExternalError(err, "setting deal info in AMT failed")
+		}
+	}
+
+	dealsCid, err := deals.Flush()
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "saving deals AMT")
+	}
+
+	self.Deals = dealsCid
+
+	nroot, err := vmctx.Storage().Put(&self)
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "storing state failed")
+	}
+
+	aerr := vmctx.Storage().Commit(old, nroot)
+	if aerr != nil {
+		return nil, aerr
+	}
+
+	return nil, nil
+}
+
+type ProcessStorageDealsPaymentParams struct {
+	DealIDs []uint64
+}
+
+func (sma StorageMarketActor) ProcessStorageDealsPayment(act *types.Actor, vmctx types.VMContext, params *ProcessStorageDealsPaymentParams) ([]byte, ActorError) {
+	var self StorageMarketState
+	old := vmctx.Storage().GetHead()
+	if err := vmctx.Storage().Get(old, &self); err != nil {
+		return nil, err
+	}
+
+	deals, err := amt.LoadAMT(types.WrapStorage(vmctx.Storage()), self.Deals)
+	if err != nil {
+		// TODO: kind of annoying that this can be caused by gas, otherwise could be fatal
+		return nil, aerrors.HandleExternalError(err, "loading deals amt")
+	}
+
+	// TODO: Would be nice if send could assert actor type
+	workerBytes, aerr := vmctx.Send(vmctx.Message().From, MAMethods.GetWorkerAddr, types.NewInt(0), nil)
+	if aerr != nil {
+		return nil, aerr
+	}
+	providerWorker, err := address.NewFromBytes(workerBytes)
+	if err != nil {
+		return nil, aerrors.HandleExternalError(err, "parsing provider worker address bytes")
+	}
+
+	for _, deal := range params.DealIDs {
+		var dealInfo OnChainDeal
+		if err := deals.Get(deal, &dealInfo); err != nil {
+			if _, is := err.(*amt.ErrNotFound); is {
+				return nil, aerrors.New(2, "deal not found")
+			}
+			return nil, aerrors.HandleExternalError(err, "getting deal info failed")
+		}
+
+		if dealInfo.Deal.Proposal.Provider != vmctx.Message().From {
+			return nil, aerrors.New(3, "ProcessStorageDealsPayment can only be called by deal provider")
+		}
+
+		if vmctx.BlockHeight() < dealInfo.ActivationEpoch {
+			// TODO: This is probably fatal
+			return nil, aerrors.New(4, "ActivationEpoch lower than block height")
+		}
+
+		if vmctx.BlockHeight() > dealInfo.ActivationEpoch+dealInfo.Deal.Proposal.Duration {
+			// Deal expired, miner should drop it
+			// TODO: process payment for the remainder of last proving period
+			return nil, nil
+		}
+
+		// todo: check math (written on a plane, also tired)
+		// TODO: division is hard, this more than likely has some off-by-one issue
+		toPay := types.BigDiv(types.BigMul(dealInfo.Deal.Proposal.StoragePrice, types.NewInt(build.ProvingPeriodDuration)), types.NewInt(dealInfo.Deal.Proposal.Duration))
```

Comment:
It would be better if `StoragePrice` was per block/per proving period. This way you don't have to divide.

Reply:
That's the plan (spec team knows), I'll handle with changing pricing to be per GB/block

Reply:
(in a followup PR)